Getting rid of heapsize step 1: miner

mocsy opened the pull request:
size_of_val in std seems to work fine.
This would allow this project to get rid of heapsize crate dep.

May or may not help out with https://github.com/paritytech/parity-ethereum/issues/9953

mocsy commented:
[clabot:check]

sorpaas commented on ethcore/transaction/tests/mod.rs:
```diff
@@ -0,0 +1,76 @@
+		assert_eq!(4, std::mem::size_of_val(&5u32));
+		assert_eq!(1, std::mem::size_of_val(&5u8));
+
+        use heapsize::HeapSizeOf;
```
Please clean this to use tab for indentation!

mocsy replied:
Done

mocsy commented:
6 is the size of the data of:
`let bytes: Vec<u8> = vec![1, 15, 31, 63, 127, 255];`
On windows both heap_size and std::mem give 6, that's fine.

Using heapsize the linux story is different:
On stable it claims data is padded to 2^n, which is 8 in this case.
On nightly reports 32bit which is wrong, since nightly no longe uses jemalloc I could actually measure using heaptrack and std::mem is spot on.

I propose to measure the data size using std::mem and remove heapsize.
I updated the test I provided accordingly.

![mem_size](https://user-images.githubusercontent.com/425404/49328481-f8dad280-f571-11e8-8258-8f769dc69a49.PNG)

Note: heap_size_of_children needs to be removed later when all packages use heap_size_of_data instead.
In this PR I updated miner only.

lazaridiscom commented:
In my attempt to understand all this, I hit on a comment which is possibly relevant:

servo/heapsize#80 (comment)

mocsy commented:
> In my attempt to understand all this, I hit on a comment which is possibly relevant:
> 
> [servo/heapsize#80](https://github.com/servo/heapsize/issues/80) (comment)

My understanding is the goal here is to generate statistics and for that size_of_val should work fine.
size_of_val tells us the size of the data, and not how many bytes the allocator would allocate.
A custom allocator could reserve more bytes than the data needs. This might not tell if the allocator uses extra spacce for something or not. The std feature is still useful to recognize if data has been freed or not, and the number would be more consistent between platforms.
The other option would be to use allocator specific functions.
Heapsize remains jemalloc specific, and malloc_size_of looks specific to me as well.

lazaridiscom commented:
I read somewhere that heapsize is more precise and faster (due to jemalloc beeing faster).

Shouldn't the goal be to remove any barriers (fix issues) that block the usage of heapsize?

Or is it save to say that "heapsize has nothing positive to offer, and should be removed" ?

(I ask very general and more out of curiosity! Please do not feel annoyed, its not about your work/PR)

mocsy commented:
What block the usage of heapsize is rustc switching migrating off of jemalloc.
heapsize remains jeamalloc only as you found https://github.com/servo/heapsize/issues/80

On https://github.com/paritytech/parity-ethereum/issues/9953 two approaches were discussed, switching back to jemalloc and keep using heapsize, or embracing the new default and stop depending on heapsize.

This PR proposes a way to do the second.

About heapsize being precise: it assumes jemalloc and it's precise as long as jemalloc is used, but if you look at the screenshot and data I posted above, it shows it's not precise with the new version of rust.

lazaridiscom commented:
Thank you for taking the time to elaborate.

I don't know the Rust Domain enough in order to know good reference projects. The standard-procedure is too look at the reference-projects, in order to see how they deal with the issue.

As an example, servo removes heapsize, and one could simply lookup the way it is replaced and adopt it:

* https://github.com/servo/webrender/pull/2192
* https://github.com/servo/servo/pull/19510
* servo/euclid@fb143ba5

So, I assume that if a heapsize-related change is similar to what the servo project (or other reference-projects) do, then it should be fine.

(clarifying again: me commenting abstract/general)

5chdn commented:
needs a 2nd reviewer, @ordian @seunlanlege anyone? :)

ordian commented:
@mocsy thanks for your contribution. The fix for #9953 is being worked on in paritytech/parity-common/pull/89.

I believe `mem::size_of_val` doesn't fully fit our purposes, e.g. it shows its size and not capacity (the actual amount of allocated memory). So I'm closing this PR.
